pkg/openwsn: re-integrate the network stack as a package (#13824)
RIOT-OS/RIOT: pkg/openwsn/doc.txt

/**
 * @defgroup pkg_openwsn   OpenWSN network stack
 * @ingroup  pkg
 * @ingroup  net
 * @brief    Provides a RIOT adaption of the OpenWSN network stack
 * @see      https://github.com/openwsn-berkeley/openwsn-fw
 *
 *
 * # OpenWSN RIOT Port
 *
 * This implementation integrates the [OpenWSN](https://github.com/openwsn-berkeley/openwsn-fw)
 * full stack (UDP, IPv6, RPL, 6TiSCH) into RIOT.
 *

>  * We provide a new RIOT "board" to the OpenWSN software. In this way we connect

Reviewer: We ? Can we rephrase the documentation to make it more impersonal ?

 * RIOT's hardware abstraction to OpenWSN's interfaces. Furthermore, we make use
 * of the simple scheduling mechanism in OpenWSN which we run in a RIOT thread
 * with second highest priority after the radio thread (THREAD_PRIORITY_MAIN - 4).
 *

>  * The current PORT of OpenWSN currently needs a root node that works along an

Reviewer: PORT ?

 * external tool that performs routing and handles join procedure:
 * [Openvisualizer](https://github.com/openwsn-berkeley/openvisualizer)
 *
 * ## Joining a network
 *
 * The first thing a new mote will need to do is to find a network. On boot it
 * will actively be listening for enhanced beacons. Once a beacon is received it
 * will adjust its timers drift and synchronize with the network. Re-synchronization
 * will be happening constantly to compensate for oscillator and timer drifts.
 *
 * Once synchronized the node will need to join the network. OpenWSN uses CoJP
 * [constrained join protocol](https://datatracker.ietf.org/doc/draft-ietf-6tisch-minimal-security/)

>  * OpenWSN stack only handles Join Requests. The JRC is not running on the root

Reviewer: JRC ? (Join Request Controller ?). Maybe give explain the acronyms, here and below (DIS, DIO, SRH, etc)

 * node, but alongside it, in the `OpenVisualizer` external tool.
 *
 * Once joined the device has the required keys to start listening to DIS and
 * to send DIO requests. Once it know about the topology of the network it will
 * be able to send packets.
 *
 * OpenWSN uses source routing. This means that unless the recipient of a packet
 * is one of the parents in the RPL tree the packet will have to go up the tree
 * upto the rootnode. But in OpenWSN RPL implementation the node does not know
 * how to route, instead it is `OpenVisualizer` which generates an SRH, attaches
 * to the incoming packet and sends it down the tree.
 *
 * ## Hardware abstraction implementation
 *
 * Following, we share some insights about the implementation of selected
 * hardware modules.
 *
 * ### sctimer
 *
 * The `sctimer` ("single compare timer") in OpenWSN is the lowest timer
 * abstraction which is used by the higher layer timer module `opentimers`. In

>  * the end t is responsible for scheduling on the MAC layer. To enable low power

Reviewer 2: ```suggestion
 * the end it is responsible for scheduling on the MAC layer. To enable low power
```

 * energy modes, this timer usually uses the RTC (real time clock) or RTT (real
 * time timer) module.
 *
 * In order to get the most portable code, this implementation uses ztimer and
 * defines a new `ztimer_clock` (ZTIMER_32768) that operates at 32768Khz to have
 * a resolution of ~30usec/tick (same as OpenWSN).
 *
 * When available ZTIMER_32768 will be built on top of `periph_rtt` to get low
 * power capabilities. If not it will be built on top of a regular timer. In
 * either case it will be shifted up if the base frequency y lower than 32768Hz
 * or frac if higher.
 *
 * The `sctimer` is responsible to set the next interrupt. Under circumstances,
 * it may happen, that the next interrupt to schedule is already late, compared
 * to the current time. In this case, timer implementations in OpenWSN directly
 * trigger a hardware interrupt. Until able to trigger sw isr directly we set
 * the callback 0 ticks in the future, which internally will be set to `now + 2`.
 *
 * ### radio
 *

>  * The radio adaptation runs in an own thread with the highest priority

Reviewer 2: ```suggestion
 * The radio adaptation runs in its own thread with the highest priority
```

 * (`THREAD_PRIORITY_MAIN - 4`) and maps to RIOT's `netdev` API.
 *
 * Hardware MAC layer features such as CSMA/CA, ACK handling and retransmissions
 * are handled by OpenWSN, so the radio driver must support disabling AUTOACK
 * and CSMA handling by the hardware. Frame filtering must as well be disabled.
 *
 * The radio adaptation preloads the buffer so `NETOPT_PRELOADING` must be
 * supported.
 *

>  * OpenWSN needs to be notified when a frame reception/transmition stats and

Reviewer 2: ```suggestion
 * OpenWSN needs to be notified when a frame reception/transmition starts and
```

 * when it ends. Therefore radio drivers need to support the following netdev
 * events:
 *
 *     - `NETDEV_EVENT_RX_STARTED`
 *     - `NETDEV_EVENT_TX_STARTED`
 *     - `NETDEV_EVENT_RX_COMPLETE`
 *     - `NETDEV_EVENT_TX_COMPLETE`
 *
 * OpenWSN expects to recover crc information on every received frame even if it
 * will simply drop frames with invalid crc. The stack can function correctly if
 * radio drivers automaticaly drop frames with an invalid crc (i.e. the stack
 * doesn't get notified about these frames), but it might print the following
 * error if using `openwsn_serial`:
 *
 * - `[IEEE802154E] wdDataDuration overflows while at state 19 in slotOffset 0`
 *
 * ### uart
 *
 * In RIOT, the first configured UART device is mapped to STDIO in most cases.
 * In OpenWSN however, the `openserial` tool uses UART to feed external software
 * running on a host computer such as
 * [Openvisualizer](https://github.com/openwsn-berkeley/openvisualizer).
 * To enable use of these tools, we provide a UART adaptation.
 *
 * By default when `openwsn_serial` (`openserial`) is used STDIO will be disabled
 * (it will use `stdio_null`). When multiple uart are available STDIO and `openserial`

>  * can be used in parallel..

Reviewer 2: ```suggestion
 * can be used in parallel.
```

 *
 * OpenWSN uart abstraction makes use of tx hardware interrupts to execute a

>  * previously registered callback after every byte is sent out. This interrupts

Reviewer 2: ```suggestion
 * previously registered callback after every byte is sent out. These interrupts
```

 * incorporate Linux WPAN tools. In addition to that, there's also other
 * hardware such as the ATUSB IEEE 802.15.4 USB Adapter which can directly be
 * used on your Linux computer with WPAN tools installed. If you conduct your
 * experiments on the IoT-LAB testbed you might want to use a:
 * [sniffer profile](https://www.iot-lab.info/tutorials/radio-sniffer).
 * - To explore the channel hopping mechanism there are rather expensive
 * multi-channel sniffers such as the BeamLogic 802.15.4 Site Analyzer that can
 * sniff all channels simultaneously. Alternatively you can set up multiple
 * separate sniffer devices locally or make use of the `sniffer_aggregator` on
 * the IoT-LAB testbed.
 * - There is a collection of external tools to interact with the IoT nodes from
 * a host computer via `openserial`. Please refer to the
 * [OpenWSN software](https://github.com/openwsn-berkeley/openwsn-sw) repository
 * for further information.
 *
 * ## Todos
 *
 * - `sctimer` to trigger an ISR immediately using sw isr.
 * - `RTT_FREQUENCY` is not configurable for most platforms, implementations
 *   should be adapted to make this configurable to be able to set RTT_FREQUENCY

>  *   to the highest possible value.

Reviewer 2: 32768 would be preferable to "highest possible value", no?

 * - The UART wrapper uses a peripheral timer to fake an interrupt after one byte
 *   has been sent. this should also be done with sw isr.
 *
 * ## Future Steps
 *
 * OpenWSN are working on refactoring their code base. As one of the outputs of
 * this modules like `cjoin`, `udp`, `coap` will become optional. Once this is
 * upstream the support for this pkg should be adapted.
 *
 * #8570 successors managed to isolate the mac layer. The required changes have

>  * not been included in this PR. Mainly because This PR intends on supporting all

Reviewer 2: ```suggestion
 * not been included in this PR. Mainly because this PR intends to support all
```

 * second highest priority after the radio thread (THREAD_PRIORITY_MAIN - 4).
 *
 * The current port of OpenWSN currently needs a root node that works along an
 * external tool that performs routing and handles join procedure:
 * [Openvisualizer](https://github.com/openwsn-berkeley/openvisualizer)
 *
 * ## Joining a network
 *
 * The first thing a new mote will need to do is to find a network. On boot it
 * will actively be listening for enhanced beacons. Once a beacon is received it
 * will adjust its timers drift and synchronize with the network. Re-synchronization
 * will be happening constantly to compensate for oscillator and timer drifts.
 *
 * Once synchronized the node will need to join the network. OpenWSN uses CoJP
 * [constrained join protocol](https://datatracker.ietf.org/doc/draft-ietf-6tisch-minimal-security/)
 * OpenWSN stack only handles Join Requests. The JRC (join registrar/coordinator,
 * a central entity) is not running on the root node, but alongside it, in the
 * `OpenVisualizer` external tool.
 *
 * Once joined the device has the required keys to start listening to DIS (DODAG
 * Information Solicitation) messages and to send DIO (DODAG Information Object)

>  * requests. Once it know about the topology of the network it will be able to

Author: fixed

 * is one of the parents in the RPL tree the packet will have to go up the tree
 * to the rootnode. But in OpenWSN RPL implementation the node does not know
 * how to route, instead it is `OpenVisualizer` which generates an SRH (Source
 * Routing Header ), attaches to the incoming packet and sends it down the tree.
 *
 * ## Hardware abstraction implementation
 *
 * Following, details about the implementation of selected hardware modules.
 *
 * ### sctimer
 *
 * The `sctimer` ("single compare timer") in OpenWSN is the lowest timer
 * abstraction which is used by the higher layer timer module `opentimers`. In
 * the end it is responsible for scheduling on the MAC layer. To enable low power
 * energy modes, this timer usually uses the RTC (real time clock) or RTT (real
 * time timer) module.
 *
 * In order to get the most portable code, this implementation uses ztimer and
 * defines a new `ztimer_clock` (ZTIMER_32768) that operates at 32768Khz to have
 * a resolution of ~30usec/tick (same as OpenWSN).
 *
 * When available ZTIMER_32768 will be built on top of `periph_rtt` to get low
 * power capabilities. If not it will be built on top of a regular timer. In

>  * either case it will be shifted up if the base frequency y lower than 32768Hz

Author: fixed

 *
 * ### radio
 *
 * The radio adaptation runs in its own thread with the highest priority
 * (`THREAD_PRIORITY_MAIN - 4`) and maps to RIOT's `netdev` API.
 *
 * Hardware MAC layer features such as CSMA/CA, ACK handling and retransmissions
 * are handled by OpenWSN, so the radio driver must support disabling AUTOACK
 * and CSMA handling by the hardware. Frame filtering must as well be disabled.
 *
 * The radio adaptation preloads the buffer so `NETOPT_PRELOADING` must be
 * supported.
 *
 * OpenWSN needs to be notified when a frame reception/transmition starts and
 * when it ends. Therefore radio drivers need to support the following netdev
 * events:
 *
 *     - `NETDEV_EVENT_RX_STARTED`
 *     - `NETDEV_EVENT_TX_STARTED`
 *     - `NETDEV_EVENT_RX_COMPLETE`
 *     - `NETDEV_EVENT_TX_COMPLETE`
 *
 * OpenWSN expects to recover crc information on every received frame even if it
 * will simply drop frames with invalid crc. The stack can function correctly if
 * radio drivers automaticaly drop frames with an invalid crc (i.e. the stack
 * doesn't get notified about these frames), but it might print the following
 * error if using `openwsn_serial`:
 *

>  * - `[IEEE802154E] wdDataDuration overflows while at state 19 in slotOffset 0`

Author: added some more details respect to this error

 *
 * ### uart
 *
 * In RIOT, the first configured UART device is mapped to STDIO in most cases.
 * In OpenWSN however, the `openserial` tool uses UART to feed external software
 * running on a host computer such as
 * [Openvisualizer](https://github.com/openwsn-berkeley/openvisualizer).
 * To enable use of these tools, a UART adaptation is provided.
 *
 * By default when `openwsn_serial` (`openserial`) is used STDIO will be disabled
 * (it will use `stdio_null`). When multiple uart are available STDIO and `openserial`

>  * can be used in parallel.

Author: I'll add something

 *
 * OpenWSN uart abstraction makes use of tx hardware interrupts to execute a
 * previously registered callback after every byte is sent out. These interrupts
 * are currently not defined in RIOT in a generic way, so instead a timer is to

>  * fire shortly after a byte is written until sw_isr are possible.

Author: done

 *
 * This implementation uses `ztimer` since its already used as an abstraction

>  * for `periph_rtt`.

Author: It explains that it doesn't use `periph_timer` directly since `ztimer` is already pulled in as a dependency for `sctimer`, I'll rephrase.

>  * for `periph_rtt`.

Author: or remove.

>  * for `periph_rtt`.

Author: Rephraased

 * OpenWSN has done that for most of their boards. These values can not be taken
 * directly from OpenWSN since they do not necessarily use the same TIMER's or
 * clock speeds.
 *
 * For more details on those parameters refer to:
 * https://openwsn.atlassian.net/wiki/spaces/OW/pages/688251/State+Machine
 *
 * Since all these parameters are HW dependant, it also means that hybrid
 * networks (different type of underlying hardware) might desynchronize often,
 * or not manage to keep in sync at all.
 *
 * ### ledpins & debugpins
 *
 * The OpenWSN software provides different hooks all over the stack to toggle
 * different LEDs as well as debug pins to examine state and scheduling of a node.
 * Default configuration files are provided for both. The LED configuration maps to
 * RIOTs `LEDX_PIN` definitions, if available. On Nucleo boards LED0 line is
 * shared SPI, so is not used.
 * The default configuration can be overwritten by setting `OPENWSN_LEDPINS_DEFAULT`
 * in the form of `ledpins_config_t`. The debugpins work similarly by setting
 * `OPENWSN_DEBUGPINS_BOARD` in the form of `debugpins_config_t`.
 *

>  * The default configuration maps to OpenWSN reference hardware `openmote-b`.

Author: Yes, I was hopping to support cc2538, by now but I still have issues. It still makes sense to me to match this configuration since the hardware has the exact amount of pins and leds.

 * incorporate Linux WPAN tools. In addition to that, there's also other
 * hardware such as the ATUSB IEEE 802.15.4 USB Adapter which can directly be
 * used on your Linux computer with WPAN tools installed. If you conduct your
 * experiments on the IoT-LAB testbed you might want to use a:
 * [sniffer profile](https://www.iot-lab.info/tutorials/radio-sniffer).
 * - To explore the channel hopping mechanism there are rather expensive
 * multi-channel sniffers such as the BeamLogic 802.15.4 Site Analyzer that can
 * sniff all channels simultaneously. Alternatively you can set up multiple
 * separate sniffer devices locally or make use of the `sniffer_aggregator` on
 * the IoT-LAB testbed.
 * - There is a collection of external tools to interact with the IoT nodes from
 * a host computer via `openserial`. Please refer to the
 * [OpenWSN software](https://github.com/openwsn-berkeley/openwsn-sw) repository
 * for further information.
 *
 * ## Todos
 *
 * - `sctimer` to trigger an ISR immediately using sw isr.
 * - `RTT_FREQUENCY` is not configurable for most platforms, implementations
 *   should be adapted to make this configurable to be able to set RTT_FREQUENCY

>  *   to 32768Hz or the closer possible value.

Author: Yes, for most platforms the clock is 32Khz, but the define `RTT_FREQUENCY` is sometimes fixed at a value, if its too small (e.g. 1Hz) then the timer will not operate adequately enough even when shifted (since this is still in a trick to fake a "faster" timer at the cost of some tick inaccuracy by loosing the amount of bits you shift). And when you scale or shift there is still some cost, whether it is tick inaccuracy or conversion time, so when possible you would still want to set it to 32768. 

I've moved some platforms to be able to configure `RTT_FREQUENCY`, e.g. `stm32` and `nrf5x`, but I haven't done it for others like `sam0` (I'm just lucky that there the fixed value is 32768).

>  * - The UART wrapper uses a peripheral timer to fake an interrupt after one byte

Author: fixed

 *   has been sent. this should also be done with sw isr.
 *
 * ## Future Steps
 *

>  * OpenWSN are working on refactoring their code base. As one of the outputs of

Author: fixed

 * this modules like `cjoin`, `udp`, `coap` will become optional. Once this is
 * upstream the support for this pkg should be adapted.
 *
 * #8570 successors managed to isolate the mac layer. The required changes have
 * not been included in this PR. Mainly because this PR intends to support all
 * network security features provided by OpenWSN. And as of the current status
 * of the upstream repository the whole stack must currently be taken.
 *
 * With the following re-works in OpenWSN this should not be necessary. Also in
 * future steps we plan to re-add the work done by @jia200x towards this.
 *
 * As more immediate future steps:
 *
 * - [ ] Follow up PR to add OpenWSN sock
 * - [ ] Complete support of OpenWSN default HW (openmote-b)

>  * - [ ] Add support in RIOT for openvisualizer

Author: Integrated in the PR!

 GNRC, provides a full 6TiSCH implementation. It does not yet support RIOT's
 sock API, so applications will have to be written against OpenWSN's API.

 This port provides a new RIOT "board" to the OpenWSN software. In this way
 RIOT's hardware abstraction connects to OpenWSN's interfaces.

 The simple scheduling mechanism in OpenWSN is run in a RIOT thread with
 second highest priority after the radio thread (THREAD_PRIORITY_MAIN - 4).

 The current port of OpenWSN currently needs a root node that works along an
 external tool that performs routing and handles join procedure:
 [Openvisualizer](https://github.com/openwsn-berkeley/openvisualizer)

 ## Joining a network

 The first thing a new mote will need to do is to find a network. On boot it
 will actively be listening for enhanced beacons. Once a beacon is received it
 will adjust its timers drift and synchronize with the network. Re-synchronization
 will be happening constantly to compensate for oscillator and timer drifts.

 Once synchronized the node will need to join the network. OpenWSN uses CoJP
 [constrained join protocol](https://datatracker.ietf.org/doc/draft-ietf-6tisch-minimal-security/)

>  OpenWSN stack only handles Join Requests. The JRC (join registrar/coordinator,

Reviewer: The second sentence is weird. Shouldn't it be "OpenWSN uses CoJP
 [constrained join protocol](https://datatracker.ietf.org/doc/draft-ietf-6tisch-minimal-security/)
 and only handles Join Requests" instead ?

 a central entity) is not running on the root node, but alongside it, in the
 `OpenVisualizer` external tool.

 Once joined the device has the required keys to start listening to DIS (DODAG
 Information Solicitation) messages and to send DIO (DODAG Information Object)

>  requests. Once it knows about the topology of the network it will be able to

Reviewer: ```suggestion
 requests. Once it knows about the topology of the network and it is be able to
```

 send packets.

 OpenWSN uses source routing. This means that unless the recipient of a packet
 is one of the parents in the RPL tree the packet will have to go up the tree
 to the root node. But in OpenWSN RPL implementation the node does not know
 how to route, instead it is `OpenVisualizer` which generates an SRH (Source

>  Routing Header ), attaches to the incoming packet and sends it down the tree.

Reviewer: ```suggestion
 Routing Header), attaches to the incoming packet and sends it down the tree.
```

 ## Hardware abstraction implementation

 Following, details about the implementation of selected hardware modules.

 ### sctimer

 The `sctimer` ("single compare timer") in OpenWSN is the lowest timer
 abstraction which is used by the higher layer timer module `opentimers`. In
 the end it is responsible for scheduling on the MAC layer. To enable low power
 energy modes, this timer usually uses the RTC (real time clock) or RTT (real
 time timer) module.

 This port has two possible implementations or sctimer, one on top of

>  periph_rtt and the another on top of ztimer, `sctimer_rtt` and

Reviewer: ```suggestion
 periph_rtt and another on top of ztimer, `sctimer_rtt` and
```

 `sctimer_ztimer` respectively. If possible `ztimer` should be preferred,
 being a virtual timer it will allow RIOT applications/modules to use the
 low level RTT timer. But ztimer (any virtual timer) has some overhead
 which can be costly depending on the different platforms used, specially
 when `openserial` is required. More on this in openserial and Known Issues.

 #### sctimer_ztimer

 In order to get the most portable code, this implementation uses
 ztimer and defines a new `ztimer_clock` (`ZTIMER_32768`) that operates
 at 32768Khz to have a resolution of ~30usec/tick (same as OpenWSN).

 When available `ZTIMER_32768` will be built on top of `periph_rtt` to
 get low power capabilities. If not it will be built on top of a
 regular timer. In either case it will be shifted up if the base
 frequency is lower than 32768Hz or frac if higher.

 When next interrupt to schedule is already late, current time,

>  Implementations in OpenWSN directly trigger a hardware interrupt.

Reviewer: ```suggestion
 implementations in OpenWSN directly trigger a hardware interrupt.
```

 Until able to trigger sw isr directly a callback is set 0 ticks in
 the future, which internally will be set to `now + RTT_MIN_OFFSET`.

 #### sctimer_rtt

 In order to reduce overhead this implementation uses bare RTT. It
 expects a RTT running at 32768Hz to have a resolution of ~30usec/tick
 (same as OpenWSN). If `RTT_FREQUENCY` is lower than 32768Hz then a
 simple time-division mechanism will be used to speed up the clock.
 This only works if `RTT_FREQUENCY` is 32768Hz/2.

 When next interrupt to schedule is already late, current time,
 Implementations in OpenWSN directly trigger a hardware interrupt.
 Until able to trigger sw isr directly a callback is set
 `RTT_MIN_OFFSET` ticks in the future.

 ### radio

 The radio adaptation runs in its own thread with the highest priority

>  (`THREAD_PRIORITY_MAIN - 4`) and maps to RIOT's `netdev` API.

Reviewer: ```suggestion
 (`THREAD_PRIORITY_MAIN - 4`) and maps to RIOT's @ref netdev API.
```

 Hardware MAC layer features such as CSMA/CA, ACK handling and retransmissions
 are handled by OpenWSN, so the radio driver must support disabling AUTOACK
 and CSMA handling by the hardware. Frame filtering must as well be disabled.

>  The radio adaptation preloads the buffer so `NETOPT_PRELOADING` must be

Reviewer: ```suggestion
 The radio adaption preloads the buffer so `NETOPT_PRELOADING` must be
```

Author: Hmm adaption is rarely used in English currently, I would prefer sticking to adaptation, same below.

Reviewer: I think both words are valid.

And also adaption seems to be more widely used in RIOT:
```
git grep adaption | wc -l
119
git grep adaptation | wc -l
14

```

Author: Sorry won't change.

 supported.

 OpenWSN needs to be notified when a frame reception/transmission starts and
 when it ends. Therefore radio drivers need to support the following netdev
 events:

     - `NETDEV_EVENT_RX_STARTED`
     - `NETDEV_EVENT_TX_STARTED`
     - `NETDEV_EVENT_RX_COMPLETE`
     - `NETDEV_EVENT_TX_COMPLETE`

 OpenWSN expects to recover crc information on every received frame even if it
 will simply drop frames with invalid crc. The stack can function correctly if
 radio drivers automatically drop frames with an invalid crc (i.e. the stack
 doesn't get notified about these frames).

 ### uart

 In RIOT, the first configured uart device is mapped to STDIO in most cases.
 In OpenWSN however, the `openserial` tool uses uart to feed external software
 running on a host computer such as
 [Openvisualizer](https://github.com/openwsn-berkeley/openvisualizer).

>  To enable use of these tools, a uart adaptation is provided.

Reviewer: ```suggestion
 To enable use of these tools, a uart adaption is provided.
```

 This is provided through the `openwsn_serial` (`openserial`) module. It
 will use the next available uart that is not used by STDIO
 (checking STDIO_UART_DEV). When multiple uart are available STDIO and
 `openserial` can be used in parallel. If `stdio_null` is used then `openserial`
 will use `STDIO_UART_DEV`, otherwise it will use the next available uart.
 e.g. If `STDIO_UART_DEV = UART_DEV(1)` `OPENWSN_UART_DEV  = UART_DEV(0)` if
 there are uarts.

 OpenWSN uart abstraction makes use of tx hardware interrupts to execute a
 previously registered callback after every byte is sent out. These interrupts
 are currently not defined in RIOT in a generic way, so instead a timer is set
 to fire shortly after a byte is written.

 It uses `ztimer` to set the timer since its already pulled in as a dependency.

 ### Openserial

 As was mentioned before any OpenWSN network will require a root node which is
 connected to an `OpenVisualizer` instance running on a host computer.

>  Interaction between `OpenVisualizer` and the open is done over serial. As

Reviewer: ```suggestion
 Interaction between `OpenVisualizer` and the root node is done over serial. As
```

- `[coap:WARNING] coapRcBadRequest(reason=OSCOAP unprotect failed: oscoapError(reason=Replay protection failed))`

The join procedure uses a replay window. If a node had already joined the network
and for some reason attempts to rejoin again, then the replay windows will need
to expire for its join request to be accepted.

The following errors are platform specific.

 - samr21-xpro issues:
      - The serial debugger hinders Openserial operation, an ftdi device must
        be used.

 - sam0 issues:
      - sam0 requires 180us busy loops every time an alarm is set or the
        counter is read. Because of the later only `sctimer_rtt` can be used
        and the max. tested baudrate for openserial is of 19200 bauds.

 - iotlab-m3 issues:
      - openserial does not work reliably over 57600 bauds or when using
        sctimer_ztimer.

 ## Todos

 - `sctimer` to trigger an ISR immediately using software interrupts.
 - `RTT_FREQUENCY` is not configurable for most platforms, implementations
   should be adapted to make this configurable to be able to set `RTT_FREQUENCY`
   to 32768Hz or the closer possible value.
 - The UART wrapper uses ztimer to fake an interrupt after one byte

>    has been sent. this should also be done with software interrupts.

Reviewer: ```suggestion
   has been sent. This should also be done with software interrupts.
```

 ##  Future Steps

>  The OpenWSN community is working working on refactoring their code base. As

Reviewer: ```suggestion
 The OpenWSN community is working on refactoring their code base. As
```

Reviewer: or maybe:
```suggestion
 The OpenWSN community is working hard on refactoring their code base. As
```
:)
